File: Beakjoon/_etc/Java/2579 계단 오르기.java
import java.util.Scanner;

public class Main {
    public static void main(String args[]) {
      Scanner s = new Scanner(System.in);
      int n = s.nextInt();
      
      int[] arr = new int[301];
      int[] dp = new int[301];
      
      for(int i = 1; i <= n; i++) {
          int num = s.nextInt();
          arr[i] = num;
      }
      dp[0] = 0;
      dp[1] = arr[1];
      dp[2] = arr[1] + arr[2];
      
      for(int i = 3; i <= n; i++) {
          dp[i] = Math.max(arr[i] + arr[i-1] + dp[i-3], arr[i] + dp[i-2]);
      }
      System.out.println(dp[n]);
      
    }
}

// http://boj.kr/5a507d1884864bcbae0e6fe50b7884df

File: Beakjoon/_etc/Kotlin/2579 계단 오르기.kt
import java.util.Scanner
import kotlin.math.*

fun main(args: Array<String>) {
    val sc: Scanner = Scanner(System.`in`)
    val num = sc.nextInt()
    
    var arr = Array(301, {0})
    var dp = Array(301, {0})
    
    for(i in 1..num) {
        val n = sc.nextInt()
        arr[i] = n;
    }
    
    dp[0] = 0
    dp[1] = arr[1]
    dp[2] = arr[1] + arr[2]
    
    for(j in 3..num) {
        dp[j] = max(arr[j] + arr[j-1] + dp[j-3], arr[j] + dp[j-2])
    }
    println("${dp[num]}")
}

// http://boj.kr/e9f0d39f974542299b73eae2ff5afce4
